### ISSUE TITLE
Отрефакторить подписки

### ISSUE THREAD

**Issue by celestora** [OpenVK/openvk#429]
На данный момент в OpenVK используется не сильно эффективная штука с подписками и прочими отношениями. Вычисление дружеских связей, заявок и стены юзера очень "дорогие" операции на данный момент из-за этого, хотя они таковыми быть не должны вообще. К тому же, текущие средства очень неоптимальные для старых процессоров Intel Atom, сильно связывают руки в возможностях улучшения подписок и групп, а также задерживают реализацию #200, #226, #426, #184, а также списков друзей. Предполагаю, данное issue является своеобразным showstopper-ом для реализации целых 5(!) задач.

Возможное (и скорее единственное) решение этой проблемы без лишних костылей это изменить подход к хранению отношений целиком. Как минимум стоит перенять структуру подписок из OpenVK Classic, слегка поменяв:

- peerType PeerA_Type
- peerType PeerB_Type
- unsigned bigint PeerA_ID
- unsigned bigint PeerB_ID
- relationType Status
- boolean PeerA_IsOrigin
- unsigned bigint (FK-> ???) List
- relationMetaData Metadata
- unsigned bigint Since
- unsigned bigint Updated

Здесь peerType это ENUM("user", "group", "chat"), relationType это SET("subscribed", "accepted", "blocked", "twinlocked", "ignored"), а relationMetaData это SET("meet_will_go", "meet_maybe", "meet_wont_go", "retard").

Первые две колонки указывают на типы пиров в отношениях (чат, группа, юзер). Вторые две это собсна айди пиров. 5-я колонка это статус. Статус может иметь такие значения: subscribed (подписка есть), accepted (заявка принята), blocked (в чс), twinlocked (покидали друг друга в чс), ignored (в игнорище (смешное слово)). 6-я колонка указывает на то, кто тут исходящий: ну то есть Status=subscribed+PeerA_IsOrigin=1, значит что PeerA подписался, если 0 то PeerB. Разумеется, эта штука учитывается и имеет смысл только если статус не accepted или twinlocked. List содержит указатель на список, Since и Updated даты изменения/создания. В Metadata хранятся дополнительные метки такие как, например, с какой вероятностью юзер пошлёпает на встречу.
Для оптимизации быстродействия пиром А будет считаться тот, у кого айди меньше. Таким образом поиск будет более быстрым, если искать по двум пирам (индекс надо будет повесить на первые 4 колонки).
Для оптимизации поиска по 1 пиру, типу, статусу и исходящести можно навесить индекс на 1, 2, 3, 4, 5 и 6 колонку. Посколько здесь нельзя предугадать айди второго пира, то неизбежена ситуация двойного запроса индексов, однако это всё равно должно быть быстрее чем JOIN, который есть сейчас.
Ещё имеет смысл повесить индексы на List и сделать индекс 1+2+3+4+5+6+8 для поиска по метаданным ещё (во встречах пригодится).

Также я считаю что такое **необходимо** сделать не прибегая к сырым запросам. Использовать нужно только Nette DBE, чтобы случайно не наговнокодить что-то на SQL намертво привязанное к MySQL 8.

P.S. ещё у меня из головы вылетело но уже лень менять весь текст, но лучше ещё впихнуть имена всех пиров и менять их по надобности, чтобы ещё ускорить поиск по друзьям/подписчикам.

**Comment by WerySkok** [comment]
В будущем нужно также отражать отношения в #61, так что данный концепт, наверное, придётся переделать с заделом под федерацию, ибо пока я не вижу её упомянутой здесь.

**Comment by celestora** [comment]
Не могли бы вы уточнить, что вы имели этим ввиду?
Я просто не разбираюсь в федилабе :D

**Comment by ghost** [comment]
Fedilab это вообще клиент для Android, не путай.

**Comment by veselcraft** [comment]
ну понятно лох мастодон ещё не постиг

**Comment by WerySkok** [comment]
> Не могли бы вы уточнить, что вы имели этим ввиду? Я просто не разбираюсь в федилабе :D

дык подписки и другие отношения с акторами на других серверах

**Comment by celestora** [comment]
> > Не могли бы вы уточнить, что вы имели этим ввиду? Я просто не разбираюсь в федилабе :D
> 
> дык подписки и другие отношения с акторами на других серверах

Нет, о чём идёт речь мне понятно, мне непонятны детали реализации ActivityPub, из-за которых нельзя любое accepted/subscribed_для_групп преобразовать в ого_подписан в активитипабе.

**Comment by WerySkok** [comment]
в данном случае не сказано как именно будут оформляться юзеры с других инстансов, потому вот и спрашиваю, что с этим будете делать

**Comment by celestora** [comment]
А есть какие-то проблемы с отслеживанием, кто подписался чтоле?

**Comment by celestora** [comment]
![image](https://user-images.githubusercontent.com/34442450/150190011-a53c0376-0113-4597-b332-e1defc9799b2.png)

Судя по всему, таковых нет, поэтому нинаю, что тут редактировать под федеративность, поэтому да, ожидаю input от вас, @WerySkok.

**Comment by WerySkok** [comment]
Ну если дополнительных действий не требуется, то ок
